pkg/steps: add new bundle source step

AlexNPavel opened the pull request:
This commit adds a new `BundleSourceStep` to be used by bundle image
builds. It creates a new image based off of the `pipeline:src` image
and uses `sed` to replace the pullspecs specified in the ci-operator
config with the in-cluster image references.

This is an early WIP and doesn't have tests yet. Those will be added soon.

/cc @petr-muller

stevekuznetsov commented on pkg/steps/bundle_source.go:
```diff
@@ -0,0 +1,121 @@
+	return handleBuild(ctx, s.buildClient, build, dry, s.artifactDir, s.dryLogger)
+}
+
+func bundleSourceDockerfile(manifestDir string, substitutions []api.PullSpecSubstitution) string {
```
At the least we need to test the output of this, at best we somehow test that the embedded commands do what we expect

AlexNPavel commented on pkg/defaults/defaults.go:
```diff
@@ -450,6 +452,21 @@ func stepConfigsForBuild(config *api.ReleaseBuildConfiguration, jobSpec *api.Job
+				Substitute:        image.Substitute,
+			}})
+			// Use the bundle source name as the "FROM" for bundle images
+			image = &api.ProjectDirectoryImageBuildStepConfiguration{
```
Yeah, that was a mistake. Fixed it.

AlexNPavel commented:
Added new tests to verify that the `find` + `sed` command works (and fixed some minor issues with it I discovered after making the tests) and that the generated dockerfile is correct.

AlexNPavel commented:
/test integration

petr-muller commented on pkg/defaults/defaults.go:
```diff
@@ -450,6 +452,15 @@ func stepConfigsForBuild(config *api.ReleaseBuildConfiguration, jobSpec *api.Job
 	for i := range config.Images {
 		image := &config.Images[i]
+		// If current image is operator bundle, add build step for its operator bundle source
+		if len(image.OperatorManifests) > 0 {
```
nit: It's a plain string so let's compare it against an empty string for readability.

petr-muller commented on pkg/steps/bundle_source.go:
```diff
@@ -0,0 +1,125 @@
+}
+
+func replaceCommand(manifestDir, pullSpec, with string) string {
+	return fmt.Sprintf("find %s -type f -exec sed -i 's?%s?stable:%s?g' {} +", manifestDir, pullSpec, with)
```
Hmm... This replaces the pullspec with `stable:with`, which doesn't seem to be enough, we need to have a pullspec (whatever we set `IMAGE_$with` to).

AlexNPavel replied:
Changed it so it uses the full registry pullspec

petr-muller commented:
Testing locally on https://github.com/petr-muller/metering-operator/tree/manifests-for-openshift-ci-index-image:
```console
$ go run ./cmd/ci-operator/ --target=[images] \
  --config ~/Temporary/DPTP-1286/petr-muller-metering-operator-manifests-for-openshift-ci-index-image.yaml \
  --git-ref petr-muller/metering-operator@manifests-for-openshift-ci-index-image
...
```

Let's inspect https://console.svc.ci.openshift.org/k8s/cluster/projects/ci-op-16k2lb12

The relevant snippet of the config:
```
images:
- dockerfile_path: Dockerfile.reporting-operator.okd
  from: base
  to: metering-reporting-operator
- dockerfile_path: Dockerfile.metering-ansible-operator
  from: ansible-operator
  to: metering-ansible-operator
- context_dir: manifests/deploy/openshift/olm/bundle
  dockerfile_path: bundle.Dockerfile
  operator_manifests: "4.6" # this is subdir of context_dir
  to: metering-operator-bundle
  substitute:
  - pullspec: quay.io/openshift/origin-metering-reporting-operator:4.6
    with: metering-reporting-operator
  - pullspec: quay.io/openshift/origin-oauth-proxy:4.6
    with: oauth-proxy
  - pullspec: quay.io/openshift/origin-hive:4.6
    with: hive
```

petr-muller commented:
Something doesn't seem quite right:

```console
$ podman create --name bundle registry.svc.ci.openshift.org/ci-op-16k2lb12/pipeline:metering-operator-bundle /bin/bash
$ podman export bundle | tar -x
$ grep metering-reporting-operator manifests/meteringoperator.v4.6.0.clusterserviceversion.yaml 
                    value: "quay.io/openshift/origin-metering-reporting-operator:4.6"
```

Seems to be not substituted in the `-sub` image already:

```console
$ podman create --name bundle-sub registry.svc.ci.openshift.org/ci-op-16k2lb12/pipeline:metering-operator-bundle-sub
$ podman export bundle-sub | tar -x
$ grep metering-reporting-operator go/src/github.com/petr-muller/metering-operator/manifests/deploy/openshift/olm/bundle/4.6/meteringoperator.v4.6.0.clusterserviceversion.yaml 
                    value: "quay.io/openshift/origin-metering-reporting-operator:4.6"
```

AlexNPavel commented:
> Seems to be not substituted in the `-sub` image already:

Should be fixed.

AlexNPavel commented:
/test integration

petr-muller commented:
/retest

AlexNPavel commented:
/test integration

AlexNPavel commented:
/retest

petr-muller commented:
/test integration

petr-muller commented on pkg/steps/bundle_source.go:
```diff
@@ -0,0 +1,152 @@
+		if err != nil {
+			return "", fmt.Errorf("failed to get replacement imagestream for image tag `%s`", sub.With)
+		}
+		dockerCommands = append(dockerCommands, fmt.Sprintf(`CMD ["bash", "-c", "%s"]`, replaceCommand(manifestDir, sub.PullSpec, replaceSpec)))
```
Err... should this be a `RUN` instead of a `CMD`? :trollface: 

I couldn't get the modified bundle, was checking all the paths, context dirs, strings... But this looks like it could be a culprit :facepalm:

AlexNPavel replied:
Yup, that's right.

Fixed that and ran a docker build locally to see if the files in the image look right. It should be fine now.
